Receive Messages Serially

eccles opened the pull request:
New functional option to add a serialHandler that reads N messages at
once and process them one at a time - serially.

    receiver := NewReceiver(, WithSerialHandler(h, 200))

where h is an instance of Handler.

Defining a SerialHandler this way disables the normal parallel
processing defined by WithHandlers().

It is highly recommended to specify RenewMessageLock if the number of
received messages is high (say > 10). One has to be sure that processing
the number of messages within 60s is achievable.


[AB#9378](https://dev.azure.com/jitsuin/0629f48c-3979-4bbc-9026-cb06b3dfd0ae/_workitems/edit/9378)

henry739 commented on azbus/receiver.go:
```diff
@@ -294,6 +311,44 @@ func (r *Receiver) receiveMessages() error {
+			// we should wrap this in a span ...
+			// XXXX: add spans using something like processMessage/r.handleReceivedMessageWithTracingContext
+			// XXXX: add logic for RenewMessageLock here perhaps
+			err = r.serialHandler.Handle(ctx, messages)
```
This seems to take a single *ReceivedMessage but seems like we're passing in []*ReceivedMessage.

wp0pw replied:
this is new Handler type defined above afaik, and this one takes a list of messages

eccles replied:
there are 2 interfaces with Handle() methods with different signatures

eccles replied:
fixed - only one handler now

eccles committed:
Receive Messages Serially

New functional option to add a serialHandler that reads N messages at
once and process them one at a time - serially.

    receiver := NewReceiver(, WithSerialHandler(h, 200))

where h is an instance of Handler.

Defining a SerialHandler this way disables the normal parallel
processing defined by WithHandlers().

It is highly recommended to specify RenewMessageLock if the number of
received messages is high (say > 10). One has to be sure that processing
the number of messages within 60s is achievable.

AB#9378

robinbryce commented on azbus/receiver.go:
```diff
@@ -294,6 +304,55 @@ func (r *Receiver) receiveMessages() error {
+					// we need a timeout per message if RenewMessageLock is disabled
+					renewCtx, renewCancel, maxDuration = r.setTimeout(ctx, r.log, msg)
+					defer renewCancel()
+					r.processMessage(renewCtx, i+1, maxDuration, msg, r.serialHandler)
```
This doesn't work for the forestrie use case. The handling of the whole batch needs to be in the application handler

eccles replied:
The Handler recives a slice of messages which it can process as it sees fit so the logic is in the application handler surely...

honourfish commented on azbus/receiver.go:
```diff
@@ -334,10 +396,17 @@ func (r *Receiver) open() error {
+			if err != nil {
+				return fmt.Errorf("failed to open handler: %w", err)
+			}
+		}
```
just return nil in this if statement at the end and get rid of the else?

eccles commented:
obsolete - azbus is moved to go-service-common
